clang/test/SemaSYCL/undefined-functor.cpp

```diff
@@ -0,0 +1,30 @@
+// RUN: %clang_cc1 -fsycl-is-device -internal-isystem %S/Inputs -sycl-std=2020 -verify -fsyntax-only %s
+// This test checks that an error is thrown when a functor without a call operator defined is used as a kernel.
+
+#include "sycl.hpp"
+
+using namespace sycl;
+queue q;
+
+struct FunctorWithoutCallOperator; // expected-note {{forward declaration of 'FunctorWithoutCallOperator'}}
+
+struct StructDefined {
+  int x;
+};
+
+int main() {
+  
+  q.submit([&](sycl::handler &cgh) {
+    // expected-error@#KernelSingleTask {{kernel parameter must be a lambda or function object}}
+    // expected-error@+2 {{invalid use of incomplete type 'FunctorWithoutCallOperator'}}
+    // expected-note@+1 {{in instantiation of function template specialization}}
+    cgh.single_task(FunctorWithoutCallOperator{});
+  });
+
+  q.submit([&](sycl::handler &cgh) {
+    // expected-error@#KernelSingleTask {{kernel parameter must be a lambda or function object}}
+    // expected-note@+1 {{in instantiation of function template specialization}}
+    cgh.single_task(StructDefined{});
```

Comment:
So the error is triggered for declarations but not definitions? Is the error valid then? Should we be diagnosing incomplete types instead?

Reply:
We are triggering the following error `kernel parameter must be a lambda or function object`, whenever a Functor (whether just declared or defined)  does not have the call operator defined. The key is "call operator" being defined in the Functor.

This error - `invalid use of incomplete type 'FunctorWithoutCallOperator'` is an existing C++ error (unrelated to SYCL) that gets triggered when we have incomplete types used in the code - `struct FunctorWithoutCallOperator` in our case.

Since `struct StructDefined` is completely defined, we don't see the error for that case.

Hope this answered your question.

Reply:
What if the call operator is simply declared, but not defined in the function object?

Reply:
>> Should we be diagnosing incomplete types instead?
>>> This error - invalid use of incomplete type 'FunctorWithoutCallOperator' is an existing C++ error (unrelated to SYCL) that gets triggered when we have incomplete types used in the code

For some reason I thought you implemented a new diagnostic talking about missing call operators. The diagnostic is for incomplete types like I thought it should be. Sorry for the confusion.

I do have a follow-up question though. It's not clear to me why the compiler crashes without your patch. Shouldn't the diagnostic for incomplete types get trigger even without your patch? It looks like your patch is just adding a second diagnostic to same site.

Reply:
Actually my bad. The diagnostic is at the single_task definition, not the instantiation. My question about the crash remains though. Can you explain why it crashes?

Reply:
> It's not clear to me why the compiler crashes without your patch. Shouldn't the diagnostic for incomplete types get triggered even without your patch? 

Prior to my patch, the compiler would crash(from an assert failure) **AND** emit the error -  `invalid use of incomplete type 'foo'` for the following code:

```
#include "CL/sycl.hpp"
struct foo;
int main() {
  sycl::queue{}.single_task(foo{});
}
```

With my patch, I am replacing the crash with the error - `kernel parameter must be a lambda or function object`, which is why we now get 2 errors for the above code.

The error - `kernel parameter must be a lambda or function object`, was already getting triggered for invalid kernel objects. In this patch, I additionally check if the named function object has a function call operator declared or defined in it and if not, I trigger the same error.

Does this help clarify your questions?

Reply:
> What if the call operator is simply declared, but not defined in the function object?

The compiler does not trigger an error if the call operator is simply declared. 
I have updated the test with two additional test cases - Functor with just the call operator declaration and one with definition.